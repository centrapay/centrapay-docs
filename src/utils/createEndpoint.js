import { HTTPSnippet } from 'httpsnippet';

const supportedClients = {
  curl: {
    targetId: 'shell',
    options: { short: true, indent: ' ', prettifyJson: true },
  },
};

function validateData(data) {
  if (typeof data !== 'object') {
    throw new Error('A data object is required');
  }

  if (data.request !== undefined && typeof data.request !== 'object') {
    throw new Error('Request must be an object');
  }

  const requiredFields = [ 'method', 'path'];
  requiredFields.forEach(field => {
    if (!(field in data)) {
      throw new Error(`Request is missing required field ${field}`);
    }
  });
}

// eslint-disable-next-line complexity
function createSnippet(data) {
  const harObject = {
    method: data.method,
    url: new URL(data.path, 'https://service.centrapay.com/').href,
    httpVersion: 'HTTP/2.0',
  };
  if (data.request?.headers) {
    harObject.headers = Object.entries(data.request.headers).map(([name, value]) => ({ name, value }));
  }
  if (data.request?.queryString) {
    harObject.queryString = Object.entries(data.request.queryString).map(([name, value]) => ({ name, value }));
  }
  if (data.request?.payload) {
    harObject.postData = {
      mimeType: 'application/json',
      text: JSON.stringify(data.request.payload)
    };
  }
  return new HTTPSnippet(harObject);
}

function createRequests(data) {
  const snippet = createSnippet(data);
  const requests = {};
  Object.entries(supportedClients).forEach(([clientId, details]) => {
    requests[clientId] = {
      lang: details.targetId,
      code: snippet.convert(details.targetId, clientId, details.options)
    };
  });
  return requests;
}

function createResponse(data) {
  return JSON.stringify(data.response, null, 2);
}

export default function createEndpoint(data) {
  validateData(data);
  return {
    path: data.path,
    method: data.method,
    requests: createRequests(data),
    response: createResponse(data),
  };
}
