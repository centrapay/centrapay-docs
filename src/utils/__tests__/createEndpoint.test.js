import createEndpoint from '../createEndpoint.js';

describe('createEndpoint', () => {
  it('creates a curl snippet with payload', () => {
    const endpoint = createEndpoint({
      method: 'POST',
      path: '/api/test',
      request: {
        headers: {
          'X-Api-Key': '<TOKEN>',
          'Content-Type': 'application/json',
        },
        payload: {
          foo: 'bar',
          bar: 'baz',
        }
      }
    });
    expect(endpoint.method).toBe('POST');
    expect(endpoint.path).toBe('/api/test');
    expect(endpoint.requests.curl.lang).toBe('shell');
    expect(endpoint.requests.curl.code).toMatchSnapshot();
  });

  it('creates a curl snippet without payload', () => {
    const endpoint = createEndpoint({
      method: 'POST',
      path: '/api/test',
      request: {
        headers: {
          'X-Api-Key': '<TOKEN>',
          'Content-Type': 'application/json',
        },
      }
    });
    expect(endpoint.requests.curl.code).toMatchSnapshot();
  });

  it('creates a curl snippet without headers or payload', () => {
    const endpoint = createEndpoint({
      method: 'POST',
      path: '/api/test',
    });
    expect(endpoint.requests.curl.code).toMatchSnapshot();
  });

  it('transforms request with query string', () => {
    const endpoint = createEndpoint({
      method: 'GET',
      path: '/example',
      request: {
        queryString: {
          foo: 'bar',
          baz: 'boo',
        }
      },
    });
    expect(endpoint.requests.curl.code).toContain('foo=bar&baz=boo');
  });

  const invalidScenarios = [
    {
      name: 'method is not defined',
      data: { path: '/example' }
    },
    {
      name: 'path is not defined',
      data: { method: 'POST' }
    },
    {
      name: 'data is empty',
      data: {}
    },
    {
      name: 'data is undefined',
      data: undefined,
    },
    {
      name: 'data is a string',
      data: ''
    },
  ];

  invalidScenarios.forEach(({ name, data }) => {
    it(name, () => expect(() => createEndpoint(data)).toThrow());
  });
});
