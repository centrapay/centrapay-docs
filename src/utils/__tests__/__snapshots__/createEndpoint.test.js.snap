// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createEndpoint > creates a snippet for a multipart/form-data request 1`] = `
"curl -X POST \\
 https://service.centrapay.com/api/test \\
 -H 'content-type: multipart/form-data' \\
 -H 'x-api-key: <TOKEN>' \\
 -F foo=bar \\
 -F file=@file.png"
`;

exports[`createEndpoint > creates a valid JSON response 1`] = `
"{
  "foo": "bar",
  "bar": "baz"
}"
`;

exports[`createEndpoint > request > creates a curl snippet with payload 1`] = `
"curl -X POST \\
 https://service.centrapay.com/api/test \\
 -H 'content-type: application/json' \\
 -H 'x-api-key: <TOKEN>' \\
 -d '
{
 "foo": "bar",
 "bar": "baz"
}
'"
`;

exports[`createEndpoint > request > creates a curl snippet without headers or payload 1`] = `
"curl -X POST \\
 https://service.centrapay.com/api/test"
`;

exports[`createEndpoint > request > creates a curl snippet without payload 1`] = `
"curl -X POST \\
 https://service.centrapay.com/api/test \\
 -H 'content-type: application/json' \\
 -H 'x-api-key: <TOKEN>'"
`;
